### PR TITLE
XArch: Trim the code block to match the actual code length

### PR DESCRIPTION
**Pull request by kunalspathak**
RyuJIT sometimes over-estimate the instruction sizes leading to allocating more code memory than actually utilized. This PR reads the actual code size value and trims the code block array so that the real code size bytes are embedded in the R2RImage.

Related: https://github.com/dotnet/runtime/issues/57368

### CODE REVIEW

**Comment by kunalspathak** [comment]
@jkotas @davidwrighton @dotnet/jit-contrib

**Comment by kunalspathak** [comment]
@mangod9

File: src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs [hunk 1]
```diff
@@ -3252,6 +3269,7 @@ private void allocMem(ref AllocMemArgs args)
                 args.coldCodeBlock = (void*)GetPin(_coldCode = new byte[args.coldCodeSize]);
                 args.coldCodeBlockRW = args.coldCodeBlock;
             }
+            allocatedHotCodeSize = args.hotCodeSize;
```
Comment (jakobbotsch) [inline src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs:3272]: Can we just use `_code.Length` instead of introducing this field?

Comment (kunalspathak) [reply src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs:3272]: Oh yes, I used it during logging but forgot to get rid of that field.

File: src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs [hunk 2]
```diff
@@ -366,6 +366,21 @@ private void CompileMethodInternal(IMethodNode methodCodeNodeNeedingCode, Method
...
+            if (codeSize != _code.Length)
+            {
+                // If the generated code is smaller than allocatedHotCodeSize, then trim the codeBlock.
+                //
+                // Currently, hot/cold splitting is not done and hence `codeSize` just includes the size of
+                // hotCode. Once hot/cold splitting is done, need to trim respective `_code` or `_coldCode`
+                // accordingly.
+                if (_compilation.Logger.IsVerbose)
+                {
+                    Log.WriteLine($"Trimming codeSize from {_code.Length} to {codeSize}, Savings: {_code.Length - codeSize}, Method: {_methodCodeNode.Method.Name}");
```
Comment (jkotas) [inline src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs:378]: We should not be logging this to the verbose log.

Comment (kunalspathak) [reply src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs:378]: Sounds good. will remove it.

File: src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs [hunk 3]
```diff
@@ -366,6 +366,21 @@ private void CompileMethodInternal(IMethodNode methodCodeNodeNeedingCode, Method
...
+                }
+                Debug.Assert(codeSize != 0);
+                Debug.Assert(_code.Length > codeSize);
+                Array.Resize(ref _code, (int)codeSize);
```
Comment (mangod9) [inline src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs:382]: wonder if there are any page alignment considerations here?

Comment (kunalspathak) [reply src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs:382]: Not really. Alignments are taken into account in `PublishCode()`.

https://github.com/dotnet/runtime/blob/e7600e00800e0ee9826ba97146f1da8a1250ccae/src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs#L378-L380

**Review by jkotas — APPROVED** [review]
Thanks

File: src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs [hunk 4]
```diff
@@ -366,6 +366,18 @@ private void CompileMethodInternal(IMethodNode methodCodeNodeNeedingCode, Method
 #endif
             }
 
+            if (codeSize != _code.Length)
```
Comment (mangod9) [inline src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs:369]: should this check be `<` instead of `!=`?

Comment (kunalspathak) [reply src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs:369]: I had that in the assert, but probably sensible to have `<` instead. I will update it.

**Comment by kunalspathak** [comment]
Also realized that this is only applicable for xarch. For arm64, we also include the size of ro data and hence checking the unused memory is not even possible.

File: src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs [hunk 5]
```diff
@@ -366,6 +366,21 @@ private void CompileMethodInternal(IMethodNode methodCodeNodeNeedingCode, Method
 #endif
             }
 
+            if (codeSize < _code.Length)
+            {
+                if (_compilation.TypeSystemContext.Target.Architecture == TargetArchitecture.X64 ||
+                    _compilation.TypeSystemContext.Target.Architecture == TargetArchitecture.X86)
+                {
+                    // For xarch, the generated code is sometimes smaller than the memory allocated.
```
Comment (jkotas) [inline src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs:374]: Is the code always of the right size outside xarch?

If yes, should the if-check above be assert instead?

**Comment by jkotas** [comment]
> Also realized that this is only applicable for xarch. For arm64, we also include the size of ro data and hence checking the unused memory is not even possible.

Mention this in the comment?

File: src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs (hunk 5 above)
Comment (kunalspathak) [reply src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs:374]: There may be cases of `codeSize != _code.Length`. We just return the native code size through `codeSize` out parameter. There is no way to know how much of the `_code.Length` was part of ro data, so there could be a mismatch.

https://github.com/dotnet/runtime/blob/c580f49f6527ac8df28876c4156345bc0a73ba64/src/coreclr/jit/emit.cpp#L5875-L5895

**Comment by kunalspathak** [comment]
> Mention this in the comment?

Sure, I have included some more explanation about armarch.

File: src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs [hunk 6]
```diff
@@ -366,6 +366,25 @@ private void CompileMethodInternal(IMethodNode methodCodeNodeNeedingCode, Method
...
+                    // In that case, trim the codeBlock to the actual value.
+                    //
+                    // For armarch, the allocation request of `hotCodeSize` also includes the roData size
+                    // while the `codeSize` returned just contains the size of the native code. As such,
+                    // there is guarantee that for armarch, (codeSize == _code.Length) is always true.
```
Comment (BruceForstall) [inline src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs:379]: This is only true for arm64. For arm32, we don't do this.

File: src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs [hunk 7]
```diff
@@ -366,6 +366,25 @@ private void CompileMethodInternal(IMethodNode methodCodeNodeNeedingCode, Method
 #endif
             }
 
+            if (codeSize < _code.Length)
+            {
+                if (_compilation.TypeSystemContext.Target.Architecture == TargetArchitecture.X64 ||
+                    _compilation.TypeSystemContext.Target.Architecture == TargetArchitecture.X86)
+                {
+                    // For xarch, the generated code is sometimes smaller than the memory allocated.
```
Comment (BruceForstall) [inline src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs:374]: I think this is also true for arm32: I think there are situations where we can over-allocate for arm32 if long branches can opportunistically be made shorter during emission, even if branch shortening didn't catch it. (There is no issue with misestimated instruction sizes except for choosing appropriate branch instruction size)

**Review by BruceForstall — CHANGES_REQUESTED** [review]
The comment at

https://github.com/dotnet/runtime/blob/16b970a0ec215e225129cd61dc7263e1a0223489/src/coreclr/inc/corjit.h#L184-L187

for compileMethod is no longer quite true, as now the returned `nativeEntry` has specific semantics... for some platforms.

That's somewhat unsatisfying.

File: src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs (hunk 7 above)
Comment (kunalspathak) [reply src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs:374]: Ok, I never collected numbers for arm32 so was not aware. I will update the condition.

**Review by mangod9 — APPROVED** [review]

**Comment by kunalspathak** [comment]
Failure related to https://github.com/dotnet/runtime/issues/61548

@BruceForstall - Do you mind taking another look?

**Review by BruceForstall — APPROVED** [review]
I'm still not fond of the fact this creates an implicit contract that isn't documented (in corjit.h), and not for all platforms. But ok...

**Comment by kunalspathak** [comment]
@BruceForstall and me discussed a possible follow-up work of implementing a new JITEE interface for `reportMem()` that will be useful for crossgen and JIT scenarios.

**Pull request merged**